File: CS/PersonalOrganizer/Common/DataModel/EntityFramework/DbRepository.cs
using System;
using System.Linq;
using System.Data;
using System.Data.Entity;
using System.Linq.Expressions;
using System.Collections.Generic;
using PersonalOrganizer.Common.Utils;
using PersonalOrganizer.Common.DataModel;
using System.Data.Entity.Validation;
using System.Data.Entity.Infrastructure;

namespace PersonalOrganizer.Common.DataModel.EntityFramework {
    public abstract class DbRepository<TEntity, TPrimaryKey, TDbContext> : DbReadOnlyRepository<TEntity, TDbContext>, IRepository<TEntity, TPrimaryKey>
        where TEntity : class
        where TDbContext : DbContext {
        Expression<Func<TEntity, TPrimaryKey>> getPrimaryKeyExpression;
        Func<TEntity, TPrimaryKey> getPrimaryKeyFunction;
        Action<TEntity, TPrimaryKey> setPrimaryKeyAction;

        public DbRepository(DbUnitOfWork<TDbContext> unitOfWork, Func<TDbContext, DbSet<TEntity>> dbSetAccessor, Expression<Func<TEntity, TPrimaryKey>> getPrimaryKeyExpression, Action<TEntity, TPrimaryKey> setPrimaryKeyAction = null)
            : base(unitOfWork, dbSetAccessor) {
            this.getPrimaryKeyExpression = getPrimaryKeyExpression;
            this.setPrimaryKeyAction = setPrimaryKeyAction;
            if(this.setPrimaryKeyAction == null)
                this.setPrimaryKeyAction = getPrimaryKeyExpression.SetValueAction().Compile();
        }
        protected virtual TEntity CreateCore() {
            TEntity newEntity = DbSet.Create();
            DbSet.Add(newEntity);
            return newEntity;
        }
        protected virtual TEntity FindCore(TPrimaryKey key) {
            return DbSet.Find(key);
        }
        protected virtual void RemoveCore(TEntity entity) {
            try {
                DbSet.Remove(entity);
            } catch(DbEntityValidationException ex) {
                throw DbExceptionsConverter.Convert(ex);
            } catch(DbUpdateException ex) {
                throw DbExceptionsConverter.Convert(ex);
            }
        }
        protected virtual TEntity ReloadCore(TEntity entity) {
            Context.Entry(entity).Reload();
            return FindCore(GetPrimaryKeyCore(entity));
        }
        protected virtual TPrimaryKey GetPrimaryKeyCore(TEntity entity) {
            if(getPrimaryKeyFunction == null)
                getPrimaryKeyFunction = getPrimaryKeyExpression.Compile();
            return getPrimaryKeyFunction(entity);
        }
        protected virtual void SetPrimaryKeyCore(TEntity entity, TPrimaryKey key) {
            setPrimaryKeyAction(entity, key);
        }
        #region IRepository
        TEntity IRepository<TEntity, TPrimaryKey>.Find(TPrimaryKey key) {
            return FindCore(key);
        }
        void IRepository<TEntity, TPrimaryKey>.Remove(TEntity entity) {
            RemoveCore(entity);
        }
        TEntity IRepository<TEntity, TPrimaryKey>.Create() {
            return CreateCore();
        }
        TEntity IRepository<TEntity, TPrimaryKey>.Reload(TEntity entity) {
            return ReloadCore(entity);
        }
        Expression<Func<TEntity, TPrimaryKey>> IRepository<TEntity, TPrimaryKey>.GetPrimaryKeyExpression {
            get { return this.getPrimaryKeyExpression; }
        }
        void IRepository<TEntity, TPrimaryKey>.SetPrimaryKey(TEntity entity, TPrimaryKey key) {
            SetPrimaryKeyCore(entity, key);
        }
        TPrimaryKey IRepository<TEntity, TPrimaryKey>.GetPrimaryKey(TEntity entity) {
            return GetPrimaryKeyCore(entity);
        }
        #endregion
    }
}

File: VB/PersonalOrganizer/Common/DataModel/EntityFramework/DbRepository.vb
Imports System
Imports System.Linq
Imports System.Data
Imports System.Data.Entity
Imports System.Linq.Expressions
Imports System.Collections.Generic
Imports Common.Utils
Imports Common.DataModel
Imports System.Data.Entity.Validation
Imports System.Data.Entity.Infrastructure
Namespace Common.DataModel.EntityFramework
    Public MustInherit Class DbRepository(Of TEntity As Class, TPrimaryKey, TDbContext As DbContext)
        Inherits DbReadOnlyRepository(Of TEntity, TDbContext)
        Implements IRepository(Of TEntity, TPrimaryKey)
        Private _getPrimaryKeyExpression As Expression(Of Func(Of TEntity, TPrimaryKey))
        Private _getPrimaryKeyFunction As Func(Of TEntity, TPrimaryKey)
        Private _setPrimaryKeyAction As Action(Of TEntity, TPrimaryKey)
        Public Sub New(ByVal unitOfWork As DbUnitOfWork(Of TDbContext), ByVal dbSetAccessor As Func(Of TDbContext, DbSet(Of TEntity)), ByVal getPrimaryKeyExpression As Expression(Of Func(Of TEntity, TPrimaryKey)), Optional ByVal setPrimaryKeyAction As Action(Of TEntity, TPrimaryKey) = Nothing)
            MyBase.New(unitOfWork, dbSetAccessor)
            Me._getPrimaryKeyExpression = getPrimaryKeyExpression
            Me._setPrimaryKeyAction = setPrimaryKeyAction
            If Me._setPrimaryKeyAction Is Nothing Then
                Me._setPrimaryKeyAction = getPrimaryKeyExpression.SetValueAction().Compile()
            End If
        End Sub
        Protected Overridable Function CreateCore() As TEntity
            Dim newEntity As TEntity = DbSet.Create()
            DbSet.Add(newEntity)
            Return newEntity
        End Function
        Protected Overridable Function FindCore(ByVal key As TPrimaryKey) As TEntity
            Return DbSet.Find(key)
        End Function
        Protected Overridable Sub RemoveCore(ByVal entity As TEntity)
            Try
                DbSet.Remove(entity)
            Catch ex As DbEntityValidationException
                Throw DbExceptionsConverter.Convert(ex)
            Catch ex As DbUpdateException
                Throw DbExceptionsConverter.Convert(ex)
            End Try
        End Sub
        Protected Overridable Function ReloadCore(ByVal entity As TEntity) As TEntity
            Context.Entry(entity).Reload()
            Return FindCore(GetPrimaryKeyCore(entity))
        End Function
        Protected Overridable Function GetPrimaryKeyCore(ByVal entity As TEntity) As TPrimaryKey
            If _getPrimaryKeyFunction Is Nothing Then
                _getPrimaryKeyFunction = _getPrimaryKeyExpression.Compile()
            End If
            Return _getPrimaryKeyFunction(entity)
        End Function
        Protected Overridable Sub SetPrimaryKeyCore(ByVal entity As TEntity, ByVal key As TPrimaryKey)
            _setPrimaryKeyAction(entity, key)
        End Sub
        Private Function Find(ByVal key As TPrimaryKey) As TEntity Implements IRepository(Of TEntity, TPrimaryKey).Find
            Return FindCore(key)
        End Function
        Private Sub Remove(ByVal entity As TEntity) Implements IRepository(Of TEntity, TPrimaryKey).Remove
            RemoveCore(entity)
        End Sub
        Private Function Create() As TEntity Implements IRepository(Of TEntity, TPrimaryKey).Create
            Return CreateCore()
        End Function
        Private Function Reload(ByVal entity As TEntity) As TEntity Implements IRepository(Of TEntity, TPrimaryKey).Reload
            Return ReloadCore(entity)
        End Function
        Private ReadOnly Property GetPrimaryKeyExpression As Expression(Of Func(Of TEntity, TPrimaryKey)) Implements IRepository(Of TEntity, TPrimaryKey).GetPrimaryKeyExpression
            Get
                Return Me._getPrimaryKeyExpression
            End Get
        End Property
        Private Sub SetPrimaryKey(ByVal entity As TEntity, ByVal key As TPrimaryKey) Implements IRepository(Of TEntity, TPrimaryKey).SetPrimaryKey
            SetPrimaryKeyCore(entity, key)
        End Sub
        Private Function GetPrimaryKey(ByVal entity As TEntity) As TPrimaryKey Implements IRepository(Of TEntity, TPrimaryKey).GetPrimaryKey
            Return GetPrimaryKeyCore(entity)
        End Function
    End Class
End Namespace
